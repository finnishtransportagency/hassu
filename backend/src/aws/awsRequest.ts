import { HttpRequest } from "@aws-sdk/protocol-http";
import { SignatureV4 } from "@aws-sdk/signature-v4";
import { Sha256 } from "@aws-crypto/sha256-browser";
import fetch from "cross-fetch";

import { defaultProvider } from "@aws-sdk/credential-provider-node";

export async function sendSignedRequest(request: HttpRequest, service: string): Promise<{ body: unknown; statusCode: number }> {
  // Sign the request
  const signer = new SignatureV4({
    credentials: defaultProvider(),
    region: "eu-west-1",
    service,
    sha256: Sha256,
  });
  const { headers, body, method, hostname, path } = await signer.sign(request);

  const response = await fetch("https://" + hostname + path, {
    headers,
    body,
    method,
  });
  const responseBody = await response.json();
  return { body: responseBody, statusCode: response.status };
}
