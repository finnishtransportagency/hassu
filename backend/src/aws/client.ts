import SSM from "aws-sdk/clients/ssm";
import SQS from "aws-sdk/clients/sqs";
import S3 from "aws-sdk/clients/s3";
import Cloudfront from "aws-sdk/clients/cloudfront";
import * as AWSXRay from "aws-xray-sdk-core";

function produce<T>(name: string, p: () => T, override = false): T {
  const key = "produce_" + name;
  if (!(globalThis as any)[key] || override) {
    const client = p();
    try {
      (globalThis as any)[key] = AWSXRay.captureAWSClient(client);
    } catch (ignore) {
      (globalThis as any)[key] = client;
    }
  }
  return (globalThis as any)[key];
}

export const getUSEast1ssm = (): SSM => produce<SSM>("ssm_us_east_1", () => new SSM({ region: "us-east-1" }));
export const getSSM = (): SSM => produce<SSM>("ssm", () => new SSM({ region: "eu-west-1" }));
export const getS3 = (): S3 => {
  if (process.env.S3_ENDPOINT) {
    return produce<S3>("s3", () => new S3({ endpoint: process.env.S3_ENDPOINT, s3ForcePathStyle: true }));
  } else {
    return produce<S3>("s3", () => new S3({ region: "eu-west-1" }));
  }
};
export const getSQS = (): SQS => {
  return produce<SQS>("sqs", () => new SQS({ region: "eu-west-1" }));
};

export const getCloudFront = (): Cloudfront => produce<Cloudfront>("cloudfront", () => new Cloudfront({ region: "us-east-1" }));
