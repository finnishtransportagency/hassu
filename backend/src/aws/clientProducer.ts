import * as AWSXRay from "aws-xray-sdk";

export function produceAWSClient<T>(name: string, p: () => T): T {
  const key = "produceAWSClient_" + name;
  if (!(globalThis as any)[key]) {
    (globalThis as any)[key] = AWSXRay.captureAWSv3Client(p() as any);
  }
  return (globalThis as any)[key];
}
