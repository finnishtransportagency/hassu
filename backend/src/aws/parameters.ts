import fetch from "cross-fetch";
import { assertIsDefined } from "../util/assertions";
import { config } from "../config";
import { ParameterNotFound, SSM } from "@aws-sdk/client-ssm";
import { defaultProvider } from "@aws-sdk/credential-provider-node";
import { log } from "../logger";
import { SuomiFiConfig } from "../suomifi/viranomaispalvelutwsinterface/suomifi";
import { PrhConfig } from "../mml/prh/prh";

interface ParameterStoreResponse {
  Parameter: ParameterData;
  ResultMetadata: never;
}

interface ParameterData {
  ARN: string;
  DataType: string;
  LastModifiedDate: string;
  Name: string;
  Selector?: string;
  SourceResult?: string;
  Type: string;
  Value: string;
  Version: number;
}

class Parameters {
  private async getParameterForEnv(paramName: string, envName?: string): Promise<string | undefined> {
    let name;
    if (envName) {
      name = "/" + envName + "/" + paramName;
    } else {
      name = paramName;
    }
    if (config.isInTest || process.env.USE_SSM) {
      const credentials = defaultProvider({ profile: "hassudev" });
      const ssm = new SSM({ region: "eu-west-1", credentials });
      try {
        const response = await ssm.getParameter({ Name: name, WithDecryption: true });
        return response.Parameter?.Value;
      } catch (e) {
        if (e instanceof ParameterNotFound) {
          return;
        }
        throw e;
      }
    } else {
      assertIsDefined(process.env.AWS_SESSION_TOKEN, "process.env.AWS_SESSION_TOKEN puuttuu!");
      const response = await fetch(
        `http://localhost:2773/systemsmanager/parameters/get/?name=${encodeURIComponent(name)}&withDecryption=true`,
        {
          headers: {
            "X-Aws-Parameters-Secrets-Token": process.env.AWS_SESSION_TOKEN,
          },
        }
      );
      if (response.ok) {
        const data = (await response.json()) as ParameterStoreResponse;
        return data.Parameter.Value;
      } else if (!envName) {
        log.error("getParameter(" + name + ") failed", { response });
      } else {
        // ei lokiteta virhettä kun ympäristökohtaista parametria ei välttämättä löydy
        log.info("getParameter(" + name + ") failed", { response });
      }
    }
    return undefined;
  }

  async getParameter(paramName: string): Promise<string | undefined> {
    let value: string | undefined;
    if (process.env.ENVIRONMENT) {
      value = await this.getParameterForEnv(paramName, process.env.ENVIRONMENT);
    }
    if (value) {
      return value;
    }
    if (process.env.INFRA_ENVIRONMENT && process.env.ENVIRONMENT !== process.env.INFRA_ENVIRONMENT) {
      value = await this.getParameterForEnv(paramName, process.env.INFRA_ENVIRONMENT);
    }
    if (value) {
      return value;
    }
    return this.getParameterForEnv(paramName);
  }

  async getRequiredInfraParameter(paramName: string): Promise<string> {
    assertIsDefined(process.env.INFRA_ENVIRONMENT, "INFRA_ENVIRONMENT pitää olla määritelty");
    let value = await this.getParameterForEnv(paramName, process.env.INFRA_ENVIRONMENT);
    if (value) {
      return value;
    }
    value = await this.getParameterForEnv(paramName);
    if (!value) {
      throw new Error(paramName + " ei löytynyt SSM:stä ympäristöstä:" + process.env.INFRA_ENVIRONMENT);
    }
    return value;
  }

  async isAsianhallintaIntegrationEnabled(): Promise<boolean> {
    return (await this.getParameter("AsianhallintaIntegrationEnabled")) === "true";
  }

  async isUspaIntegrationEnabled(): Promise<boolean> {
    return (await this.getParameter("UspaIntegrationEnabled")) === "true";
  }

  async getUspaBaseUrl(): Promise<string | undefined> {
    return await this.getParameter("UspaBaseUrl");
  }

  async getAshaBaseUrl(): Promise<string | undefined> {
    return await this.getParameter("AshaBaseUrl");
  }

  async getAsianhallintaSQSUrl() {
    // Tuki asianhallinnan käynnistämiseen testilinkillä [oid].dev.ts kautta. Ei tarvita kun asianhallintaintegraatio on automaattisesti käytössä.
    return this.getParamOrVariable("outputs/AsianhallintaSQSUrl", "ASIANHALLINTA_SQS_URL");
  }

  async getKiinteistoSQSUrl() {
    return this.getParameter("outputs/KiinteistoSQSUrl");
  }

  async getKeycloakDomain() {
    return this.getRequiredAccountParameter("KeycloakDomain");
  }

  async getKeycloakPrivateDomain() {
    return this.getRequiredAccountParameter("KeycloakPrivateDomain");
  }

  async getKeycloakClientId() {
    return this.getRequiredAccountParameter("KeycloakClientId");
  }

  async getKeycloakUsername() {
    return this.getRequiredAccountParameter("KeycloakUsername");
  }

  async getKeycloakPassword() {
    return this.getRequiredAccountParameter("KeycloakPassword");
  }

  async isSuomiFiIntegrationEnabled(): Promise<boolean> {
    return (await this.getParameter("SuomiFiIntegrationEnabled")) === "true";
  }

  async isSuomiFiViestitIntegrationEnabled(): Promise<boolean> {
    return (await this.getParameter("SuomiFiViestitIntegrationEnabled")) === "true";
  }

  async getIndexerSQSUrl() {
    return this.getParamOrVariable("outputs/IndexerSQSUrl");
  }

  async getOmistajaIndexerSQSUrl() {
    return this.getParamOrVariable("outputs/OmistajaIndexerSQSUrl");
  }

  async getMuistuttajaIndexerSQSUrl() {
    return this.getParamOrVariable("outputs/MuistuttajaIndexerSQSUrl");
  }

  async getSuomiFiConfig() {
    const param = await this.getRequiredAccountParameter("SuomiFiConfig");
    const cfg: Partial<SuomiFiConfig> = {};
    param.split("\n").forEach((e) => {
      const v = e.split("=");
      cfg[v[0] as keyof SuomiFiConfig] = v[1].trim();
    });
    return cfg as SuomiFiConfig;
  }

  async getSuomiFiCertificate() {
    return this.getParameterForEnv("SuomiFiCertificate");
  }

  async getSuomiFiPrivateKey() {
    return this.getParameterForEnv("SuomiFiPrivateKey");
  }

  async getSuomiFiSQSUrl() {
    return this.getParameter("outputs/SuomiFiSQSUrl");
  }

  async getIsEvkActivated() {
    return this.getRequiredParameter("EvkActivationDate");
  };

  getMmlApiKey() {
    return this.getRequiredParameter("MmlApiKey");
  }

  getKtjBaseUrl() {
    return this.getRequiredParameter("KtjBaseUrl");
  }

  getOgcBaseUrl() {
    return this.getRequiredParameter("OgcBaseUrl");
  }

  getOgcApiKey() {
    return this.getRequiredParameter("OgcApiKey");
  }

  getOgcApiExmaples() {
    return this.getRequiredParameter("OgcApiExamples");
  }

  async getPrhConfig() {
    const param = await this.getRequiredParameter("PrhConfig");
    const cfg: Partial<PrhConfig> = {};
    param.split("\n").forEach((e) => {
      const v = e.split("=");
      cfg[v[0] as keyof PrhConfig] = v[1].trim();
    });
    return cfg as PrhConfig;
  }

  private async getRequiredAccountParameter(paramName: string): Promise<string> {
    const value = await this.getParameterForEnv(paramName);
    if (!value) {
      throw new Error(paramName + " ei löytynyt SSM:stä");
    }
    return value;
  }

  private async getRequiredParameter(paramName: string): Promise<string> {
    const value = await this.getParameter(paramName);
    if (!value) {
      throw new Error(paramName + " ei löytynyt SSM:stä");
    }
    return value;
  }

  private getParamOrVariable(ssmPath: string, envVariableName?: string) {
    const url = envVariableName && process.env[envVariableName];
    if (url) {
      return url;
    }
    return this.getParameter(ssmPath);
  }
}

export const parameters = new Parameters();
