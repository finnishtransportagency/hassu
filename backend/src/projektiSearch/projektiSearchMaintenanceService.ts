import { openSearchClientIlmoitustauluSyote, openSearchClientJulkinen, openSearchClientYllapito } from "./openSearchClient";
import { projektiDatabase } from "../database/projektiDatabase";
import { projektiSearchService } from "./projektiSearchService";
import projektiSettings from "./projekti-settings.json";
import projektiMapping from "./projekti-mapping.json";
import projektiJulkinenMapping from "./projekti-julkinen-mapping.json";
import { log } from "../logger";
import { DBProjekti } from "../database/model";

export type MaintenanceEvent = {
  action: "deleteIndex" | "index";
  startKey?: string;
  index?: number;
  size?: number;
  projekti?: DBProjekti;
};

export class ProjektiSearchMaintenanceService {
  async deleteIndex(): Promise<void> {
    log.info(await openSearchClientYllapito.deleteIndex());
    log.info(await openSearchClientYllapito.put("", "{}"));
    log.info(await openSearchClientYllapito.putSettings(JSON.stringify(projektiSettings)));
    log.info(await openSearchClientYllapito.putMapping(JSON.stringify(projektiMapping)));

    for (const client of Object.values(openSearchClientJulkinen)) {
      log.info(await client.deleteIndex());
      log.info(await client.put("", "{}"));
      log.info(await client.putSettings(JSON.stringify(projektiSettings)));
      log.info(await client.putMapping(JSON.stringify(projektiJulkinenMapping)));
    }

    log.info(await openSearchClientIlmoitustauluSyote.deleteIndex());
  }

  async index(event: MaintenanceEvent): Promise<string | undefined> {
    const scanResult = await projektiDatabase.scanProjektit(event.startKey);
    for (const projekti of scanResult.projektis) {
      try {
        await projektiSearchService.indexProjekti(projekti);
      } catch (e) {
        log.error(e);
      }
    }
    return scanResult.startKey;
  }
}
