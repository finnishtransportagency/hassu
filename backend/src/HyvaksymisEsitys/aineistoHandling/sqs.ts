import { SQS, SendMessageRequest } from "@aws-sdk/client-sqs";
import { produce } from "../../aws/produce";
import { log } from "../../logger";
import { SqsEvent } from "./sqsEvent";

export const getSQS = (): SQS => {
  return produce<SQS>("sqs-hyvaksymisesitys", () => new SQS({ region: "eu-west-1" }));
};

function createMessageParams(params: SqsEvent, retry?: boolean) {
  if (retry) {
    let retriesLeft = params.retriesLeft;
    if (retriesLeft == undefined) {
      retriesLeft = 60;
    } else if (retriesLeft <= 0) {
      log.error("Giving up retrying", { params });
      return;
    } else {
      retriesLeft--;
    }
    params.retriesLeft = retriesLeft;
  }
  const messageParams: SendMessageRequest = {
    MessageGroupId: params.oid,
    MessageBody: JSON.stringify({ timestamp: Date.now(), ...params }),
    QueueUrl: "", //TODO: config.hyvaksymisEsitysSqsUrl,
    DelaySeconds: retry ? 60 : undefined,
  };
  return messageParams;
}

export async function addEventToSqsQueue(params: SqsEvent, retry?: boolean) {
  const messageParams = createMessageParams(params, retry);
  if (messageParams) {
    log.info("addEventToSqsQueue", { messageParams });
    const result = await getSQS().sendMessage(messageParams);
    log.info("addEventToSqsQueue", { result });
  }
}
