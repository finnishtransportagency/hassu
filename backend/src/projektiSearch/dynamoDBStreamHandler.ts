import { log } from "../logger";
import { DynamoDBRecord, DynamoDBStreamEvent } from "aws-lambda/trigger/dynamodb-stream";
import { DBProjekti } from "../database/model";
import { projektiSearchService } from "./projektiSearchService";
import { setupLambdaMonitoring, setupLambdaMonitoringMetaData, wrapXRayAsync } from "../aws/monitoring";
import { MaintenanceEvent, ProjektiSearchMaintenanceService } from "./projektiSearchMaintenanceService";

import { unmarshall } from "@aws-sdk/util-dynamodb";
import { AttributeValue } from "@aws-sdk/client-dynamodb";
import { getSQS } from "../aws/clients/getSQS";
import { parameters } from "../aws/parameters";
import { SQSEvent } from "aws-lambda/trigger/sqs";
import { projektiDatabase } from "../database/projektiDatabase";
import { SendMessageCommandInput } from "@aws-sdk/client-sqs";

async function handleUpdate(record: DynamoDBRecord) {
  if (record.dynamodb?.NewImage) {
    const projekti = unmarshall(record.dynamodb.NewImage as unknown as Record<string, AttributeValue>) as DBProjekti;
    log.info(`${record.eventName}`, { oid: projekti.oid });

    await projektiSearchService.indexProjekti(projekti);
  } else {
    log.error("No DynamoDB record to update");
  }
}

async function handleRemove(record: DynamoDBRecord) {
  if (record.dynamodb?.Keys?.oid.S) {
    const oid: string = record.dynamodb.Keys.oid.S;
    log.info("REMOVE", { oid });
    await projektiSearchService.removeProjekti(oid);
  } else {
    log.error("No DynamoDB key to remove");
  }
}

async function handleManagementAction(event: MaintenanceEvent) {
  log.info("handleManagementAction");
  const action = event.action;
  if (action == "deleteIndex") {
    await new ProjektiSearchMaintenanceService().deleteIndex();
  } else if (action == "index") {
    let startKey: string | undefined = undefined;
    const queueUrl = await parameters.getIndexerSQSUrl();
    do {
      const scanResult = await projektiDatabase.scanProjektit(startKey);
      startKey = scanResult.startKey;
      const entries = scanResult.projektis.map<SendMessageCommandInput>((projekti) => ({
        QueueUrl: queueUrl,
        MessageBody: JSON.stringify({ action: "index", oid: projekti.oid }),
      }));
      for (const args of entries) {
        await getSQS().sendMessage(args);
      }
    } while (startKey);
    log.info("Indeksointi aloitettu");
  }
}

type Event = DynamoDBStreamEvent | MaintenanceEvent | SQSEvent;

function eventIsMaintenanceEvent(event: Event): event is MaintenanceEvent {
  return !!(event as MaintenanceEvent).action;
}

function eventIsSqsEvent(event: Event): event is SQSEvent {
  const records = (event as SQSEvent).Records;
  return !!records.length && records.every((record) => record.body);
}

export const handleDynamoDBEvents = async (event: Event): Promise<void> => {
  if (eventIsMaintenanceEvent(event)) {
    await handleManagementAction(event);
  } else if (eventIsSqsEvent(event)) {
    await handleSqsEvent(event)
  } else {
    await handleStreamEvent(event);
  }
}

async function handleSqsEvent(event: SQSEvent) {
  log.info("handleSqsEvent");
  log.info("SQS records: " + event.Records.length);
  await Promise.all(
    event.Records.map(async (record) => {
      if (!record.body) {
        return;
      }
      const body: MaintenanceEvent = JSON.parse(record.body);
      log.info("Projektin oid " + body.oid);
      if (body.oid) {
        const dbProjekti = await projektiDatabase.loadProjektiByOid(body.oid);
        if (dbProjekti) {
          await projektiSearchService.indexProjekti(dbProjekti);
        }
      }
    })
  );
  log.info("Batchin indeksointi valmis");
}

function handleStreamEvent(event: DynamoDBStreamEvent) {
  log.info("handleStreamEvent");
  setupLambdaMonitoring();
  if (!event.Records) {
    log.warn("No records");
    return;
  }
  return wrapXRayAsync("handler", async (subsegment) => {
    return (async () => {
      setupLambdaMonitoringMetaData(subsegment);
      try {
        for (const record of event.Records) {
          switch (record.eventName) {
            case "INSERT":
            case "MODIFY":
              await handleUpdate(record);
              break;
            case "REMOVE":
              await handleRemove(record);
          }
        }
      } catch (e: unknown) {
        log.error(e);
        throw e;
      }
    })();
  });
};
