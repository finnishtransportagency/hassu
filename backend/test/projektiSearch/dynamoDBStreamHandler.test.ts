import { describe, it } from "mocha";
import { handleDynamoDBEvents } from "../../src/projektiSearch/dynamoDBStreamHandler";
import { ProjektiSearchFixture } from "./projektiSearchFixture";
import {
  OpenSearchClient,
  openSearchClientIlmoitustauluSyote,
  openSearchClientJulkinen,
  openSearchClientYllapito,
} from "../../src/projektiSearch/openSearchClient";
import { ProjektiFixture } from "../fixture/projektiFixture";
import sinon, { SinonStubbedInstance } from "sinon";
import { expect } from "chai";
import { SQSEvent, SQSRecord } from "aws-lambda/trigger/sqs";
import { parameters } from "../../src/aws/parameters";
import { mockClient } from "aws-sdk-client-mock";
import { SQSClient } from "@aws-sdk/client-sqs";
import { projektiDatabase } from "../../src/database/projektiDatabase";
import { MaintenanceEvent } from "../../src/projektiSearch/projektiSearchMaintenanceService";

describe("dynamoDBStreamHandler", () => {
  let fixture: ProjektiSearchFixture;
  const projektiFixture: ProjektiFixture = new ProjektiFixture();
  const projekti1 = projektiFixture.dbProjekti1();
  const projekti2 = projektiFixture.dbProjekti2();
  const projekti3 = projektiFixture.dbProjekti5();
  let indexProjektiStub: sinon.SinonStub;
  let removeProjektiStub: sinon.SinonStub;
  let indexProjektiSuomiStub: sinon.SinonStub;
  let indexProjektiRuotsiStub: sinon.SinonStub;
  let removeProjektiSuomiStub: sinon.SinonStub;
  let removeProjektiRuotsiStub: sinon.SinonStub;
  let openSearchClientIlmoitustauluSyoteStub: SinonStubbedInstance<OpenSearchClient>;

  before(() => {
    fixture = new ProjektiSearchFixture();

    indexProjektiStub = sinon.stub(openSearchClientYllapito, "putDocument");
    removeProjektiStub = sinon.stub(openSearchClientYllapito, "deleteDocument");
    indexProjektiSuomiStub = sinon.stub(openSearchClientJulkinen["SUOMI"], "putDocument");
    indexProjektiRuotsiStub = sinon.stub(openSearchClientJulkinen["RUOTSI"], "putDocument");
    removeProjektiSuomiStub = sinon.stub(openSearchClientJulkinen["SUOMI"], "deleteDocument");
    removeProjektiRuotsiStub = sinon.stub(openSearchClientJulkinen["RUOTSI"], "deleteDocument");
    openSearchClientIlmoitustauluSyoteStub = sinon.stub(openSearchClientIlmoitustauluSyote);
    sinon.stub(parameters, "isAsianhallintaIntegrationEnabled").returns(Promise.resolve(false));
  });

  beforeEach(() => {
    openSearchClientIlmoitustauluSyoteStub.putDocument.resolves();
  });

  afterEach(() => {
    sinon.reset();
  });

  after(() => {
    sinon.restore();
  });

  it("should index new projektis successfully", async () => {
    await handleDynamoDBEvents(fixture.createNewProjektiEvent(projekti2));
    expect(indexProjektiStub.calledOnce).to.be.true;
    expect(indexProjektiStub.getCall(0).args[0]).to.be.equal(projekti2.oid);
    expect(indexProjektiStub.getCall(0).args[1]).toMatchSnapshot();
    expect(indexProjektiSuomiStub.getCall(0).args[1]).toMatchSnapshot();
    expect(indexProjektiRuotsiStub.getCall(0).args[1]).toMatchSnapshot();
    expect(openSearchClientIlmoitustauluSyoteStub.putDocument.getCalls()).to.have.length(2);
  });

  it("should index projekti updates successfully", async () => {
    await handleDynamoDBEvents(fixture.createUpdateProjektiEvent(projekti2));
    expect(indexProjektiStub.calledOnce).to.be.true;
    expect(indexProjektiStub.getCall(0).args[0]).to.be.equal(projekti2.oid);
    expect(indexProjektiStub.getCall(0).args[1]).toMatchSnapshot();
    expect(openSearchClientIlmoitustauluSyoteStub.putDocument.getCalls()).to.have.length(2);
    expect(openSearchClientIlmoitustauluSyoteStub.putDocument.getCalls().map((call) => call.args)).toMatchSnapshot();
  });

  it("should remove deleted projekti from index successfully", async () => {
    const indexedKuulutusId = "indexed_kuulutus_id";
    openSearchClientIlmoitustauluSyoteStub.query.resolves({
      hits: {
        hits: [
          {
            _id: indexedKuulutusId,
          },
        ],
      },
    });

    await handleDynamoDBEvents(fixture.createdDeleteProjektiEvent(projekti2.oid));
    expect(removeProjektiStub.calledOnce).to.be.true;
    expect(removeProjektiStub.getCall(0).firstArg).to.be.equal(projekti2.oid);
    expect(removeProjektiSuomiStub.getCall(0).firstArg).to.be.equal(projekti2.oid);
    expect(removeProjektiRuotsiStub.getCall(0).firstArg).to.be.equal(projekti2.oid);
    expect(openSearchClientIlmoitustauluSyoteStub.deleteDocument.getCalls()).to.have.length(1);
    expect(openSearchClientIlmoitustauluSyoteStub.deleteDocument.getCall(0).lastArg).to.eq(indexedKuulutusId);
  });

  it("should reindex the database successfully", async () => {
    const sqsEvent: SQSEvent = {
      Records: [
        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
        {
          messageId: "1",
          body: JSON.stringify({ action: "index", projekti: projekti2 }),
        } as SQSRecord,
      ],
    };
    await handleDynamoDBEvents(sqsEvent);
    expect(indexProjektiStub.calledOnce).to.be.true;
    expect(indexProjektiStub.getCall(0).args[0]).to.be.equal(projekti2.oid);
    expect(indexProjektiStub.getCall(0).args[1]).toMatchSnapshot();
    expect(openSearchClientIlmoitustauluSyoteStub.putDocument.getCalls()).to.have.length(2);
    expect(openSearchClientIlmoitustauluSyoteStub.putDocument.getCalls().map((call) => call.args)).toMatchSnapshot();
  });

  it("should reindex the database successfully on maintenance event", async () => {
    sinon
      .stub(projektiDatabase, "scanProjektit")
      .onFirstCall().resolves({ projektis: [projekti1, projekti2], startKey: 'startkey1' })
      .onSecondCall().resolves({ projektis: [projekti3], startKey: undefined });
    sinon.stub(parameters, "getIndexerSQSUrl").resolves("mockedIndexerSQSUrl");
    const sqsStub = mockClient(SQSClient);

    const mEvent: MaintenanceEvent = {
      action: "index",
    };
    await handleDynamoDBEvents(mEvent);
    expect(indexProjektiStub.calledOnce).to.be.false;

    expect(sqsStub.send).to.have.been.calledThrice;
    expect(sqsStub.send.args[0][0].input).toMatchSnapshot();
    expect(sqsStub.send.args[1][0].input).toMatchSnapshot();
    expect(sqsStub.send.args[2][0].input).toMatchSnapshot();
  });
});
