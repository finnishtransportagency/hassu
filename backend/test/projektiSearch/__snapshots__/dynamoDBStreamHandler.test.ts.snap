// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynamoDBStreamHandler should index new projektis successfully 1`] = `
Object {
  "asiatunnus": "A2",
  "maakunnat": Array [
    "Uusimaa",
    "Pirkanmaa",
  ],
  "muokkaajat": Array [
    "A123",
    "A000111",
  ],
  "nimi": "Testiprojekti 2",
  "paivitetty": "2022-03-15T14:30:00.000Z",
  "projektiTyyppi": "TIE",
  "projektipaallikko": "Projari, Pekka",
  "suunnittelustaVastaavaViranomainen": "UUDENMAAN_ELY",
  "vaihe": "SUUNNITTELU",
  "vaylamuoto": Array [
    "tie",
  ],
}
`;

exports[`dynamoDBStreamHandler should index new projektis successfully 2`] = `
Object {
  "hankkeenKuvaus": "Suunnittelu kohde on osa Hyvinkää-Hanko sähköistyshanketta, jossa toteutetaan myös tasoristeysten toimenpiteitä. Hyväksytyssä ratasuunnitelmassa kyseisessä kohdassa on hyväksytty suunnitelmaratkaisuna Kisan seisakkeen ja Leksvallin tasoristeysten poistaminen parantamalla Helmströmin tasoristeyksen kohdalle uusi tasoristeys. Nyt käynnistetään Rata-suunnitelman päivitys kyseisessä kohdassa ja suunnitellaan kaikkien kolmen tasoristeyksen poistaminen uudella Leksvallin ylikulkusillalla.",
  "kunnat": Array [
    "Mikkeli",
    "Juva",
    "Savonlinna",
  ],
  "maakunnat": Array [
    "Uusimaa",
    "Pirkanmaa",
  ],
  "nimi": "Marikan testiprojekti",
  "paivitetty": "2022-03-15T14:30:00.000Z",
  "projektiTyyppi": "TIE",
  "publishTimestamp": "2022-03-28T14:28:00+03:00",
  "vaihe": "ALOITUSKUULUTUS",
  "vaylamuoto": Array [
    "tie",
  ],
}
`;

exports[`dynamoDBStreamHandler should index new projektis successfully 3`] = `
Object {
  "hankkeenKuvaus": "Designplatsen är en del av elektrifieringsprojektet Hyvinge-Hangö, som också kommer att genomföra plankorsningsåtgärder. I den fastställda spårplanen är borttagandet av tävlingsstoppet och Leksvallskorsningen genom förbättring av en ny plankorsning vid Helmströms plankorsning vid denna punkt godkänd som planlösning. En uppdatering av Spårplanen kommer nu att lanseras då och man planerar att ta bort alla tre plankorsningarna på den nya Leksvallsöverfarten.",
  "kunnat": Array [
    "Mikkeli",
    "Juva",
    "Savonlinna",
  ],
  "maakunnat": Array [
    "Uusimaa",
    "Pirkanmaa",
  ],
  "nimi": "Marikas testprojekt",
  "paivitetty": "2022-03-15T14:30:00.000Z",
  "projektiTyyppi": "TIE",
  "publishTimestamp": "2022-03-28T14:28:00+03:00",
  "vaihe": "ALOITUSKUULUTUS",
  "vaylamuoto": Array [
    "tie",
  ],
}
`;

exports[`dynamoDBStreamHandler should index projekti updates successfully 1`] = `
Object {
  "asiatunnus": "A2",
  "maakunnat": Array [
    "Uusimaa",
    "Pirkanmaa",
  ],
  "muokkaajat": Array [
    "A123",
    "A000111",
  ],
  "nimi": "Testiprojekti 2",
  "paivitetty": "2022-03-15T14:30:00.000Z",
  "projektiTyyppi": "TIE",
  "projektipaallikko": "Projari, Pekka",
  "suunnittelustaVastaavaViranomainen": "UUDENMAAN_ELY",
  "vaihe": "SUUNNITTELU",
  "vaylamuoto": Array [
    "tie",
  ],
}
`;

exports[`dynamoDBStreamHandler should index projekti updates successfully 2`] = `
Array [
  Array [
    "2_aloitusKuulutus_SUOMI_2022-03-28T14:28",
    Object {
      "date": "2022-03-28T14:28:00+03:00",
      "kieli": "SUOMI",
      "kunnat": Array [
        "Mikkeli",
        "Juva",
        "Savonlinna",
      ],
      "maakunnat": Array [
        "Uusimaa",
        "Pirkanmaa",
      ],
      "oid": "2",
      "title": "Kuulutus suunnittelun aloittamisesta: Marikan testiprojekti",
      "type": "Kuulutus",
      "url": "https://localhost:3000/suunnitelma/2/aloituskuulutus",
      "vaylamuoto": Array [
        "tie",
      ],
    },
  ],
  Array [
    "2_aloitusKuulutus_RUOTSI_2022-03-28T14:28",
    Object {
      "date": "2022-03-28T14:28:00+03:00",
      "kieli": "RUOTSI",
      "kunnat": Array [
        "Mikkeli",
        "Juva",
        "Savonlinna",
      ],
      "maakunnat": Array [
        "Uusimaa",
        "Pirkanmaa",
      ],
      "oid": "2",
      "title": "RUOTSIKSI Kuulutus suunnittelun aloittamisesta: Marikas testprojekt",
      "type": "Kuulutus",
      "url": "https://localhost:3000/suunnitelma/2/aloituskuulutus",
      "vaylamuoto": Array [
        "tie",
      ],
    },
  ],
]
`;
